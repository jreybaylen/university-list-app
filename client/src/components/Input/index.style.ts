import { StypesProps } from '@interface/style.interface'

const styles: StypesProps = {
    input: {
        outline: 0,
        width: '100%',
        borderRadius: 4,
        padding: '10px 15px',
        border: '1px solid #ccc',
        textTransform: 'capitalize'
    }
}

export {
    styles
}