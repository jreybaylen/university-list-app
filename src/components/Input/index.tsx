import { ChangeEvent, InputHTMLAttributes } from 'react'

import { styles } from './index.style'

function Input (props: InputHTMLAttributes<HTMLInputElement>): JSX.Element {
    const { onChange, ...rest } = props
    const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
        if (props.onChange) {
            props.onChange(event)
        }
    }
    const inputElement = (
        <input style={ styles.input } onChange={ handleChange } { ...rest } />
    )

    return inputElement
}

export {
    Input
}