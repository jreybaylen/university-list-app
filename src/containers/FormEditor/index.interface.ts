import { FormEvent } from 'react'

interface FormEditorProps {
    title: string
    submitLabel: string
    children: Array<JSX.Element>
    onSubmit(param: FormEvent<HTMLFormElement>): void
}

export type {
    FormEditorProps
}