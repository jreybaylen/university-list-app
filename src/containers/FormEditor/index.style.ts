import { StypesProps } from '@interface/style.interface'

const styles: StypesProps = {
    container: {
        padding: 30,
        width: '90%',
        maxWidth: 400,
        borderRadius: 5,
        margin: '10% auto',
        boxShadow: 'rgb(0 0 0 / 20%) 0px 10px 15px 0px'
    },
    heading: {
        fontWeight: 500,
        marginBottom: 30
    },
    formGroup: {
        marginBottom: 15
    },
    button: {
        width: '100%',
        color: '#fff',
        border: 'none',
        borderRadius: 25,
        background: '#1a73e8'
    }
}

export {
    styles
}