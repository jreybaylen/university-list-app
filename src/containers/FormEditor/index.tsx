import { styles } from './index.style'
import { FormEditorProps } from './index.interface'

import { Form, Button } from '@components/index'

function FormEditor (props: FormEditorProps): JSX.Element {
    const handleSubmit = () => {
        if (props.onSubmit) {
            props.onSubmit()
        }
    }
    const authElement = (
        <div style={ styles.container }>
            <Form onSubmit={ handleSubmit }>
                <h1 style={ styles.heading }>
                    { props.title }
                </h1>
                { props.children.map(
                    (children: JSX.Element) => (
                        <div key={ children.key } style={ styles.formGroup }>
                            { children }
                        </div>
                    )
                ) }
                <div style={ styles.formGroup }>
                    <Button
                        type="submit"
                        style={ styles.button }
                        children={ props.submitLabel }
                    />
                </div>
            </Form>
        </div>
    )

    return authElement
}

export default FormEditor