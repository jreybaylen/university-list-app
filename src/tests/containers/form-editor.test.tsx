import { render, fireEvent } from '@testing-library/react'

import { Input } from '@components/Input'
import FormEditor from '@container/FormEditor'

const title = 'Test Form'
const formName = 'TestEditor'
const submitLabel = 'Test Button'

describe('<FormEditor /> Component', () => {
    it('Should render without crashing', () => {
        const handleSubmit = jest.fn()

        render(
            <FormEditor
                title={ title }
                children={ [] }
                name={ formName }
                onSubmit={ handleSubmit }
                submitLabel={ submitLabel }
            />
        )
    })

    it('Should render without Form, Heading, Children and Button component', () => {
        const handleSubmit = jest.fn()
        const placeholder = 'Test Input'
        const { container, getByPlaceholderText } = render(
            <FormEditor
                title={ title }
                name={ formName }
                onSubmit={ handleSubmit }
                submitLabel={ submitLabel }
            >
                <Input key="username" type="text" placeholder={ placeholder } />
                <Input key="password" type="password" placeholder="Password" />
            </FormEditor>
        )

        expect(container.querySelector('form')).toBeInTheDocument()
        expect(getByPlaceholderText(placeholder)).toBeInTheDocument()
        expect(container.querySelector('h1')?.textContent).toStrictEqual(title)
        expect(container.querySelector('button[type="submit"]')?.textContent).toStrictEqual(submitLabel)
    })

    it('Should trigger onSubmit once the form has been submitted', () => {
        const handleSubmit = jest.fn()
        const { container, getByText } = render(
            <FormEditor
                title={ title }
                name={ formName }
                children={ [] }
                onSubmit={ handleSubmit }
                submitLabel={ submitLabel }
            />
        )

        expect(handleSubmit).not.toHaveBeenCalled()
        expect(getByText(submitLabel)).toBeInTheDocument()
        expect(container.querySelector('form')).toBeInTheDocument()
        fireEvent.submit(getByText(submitLabel))
        expect(handleSubmit).toHaveBeenCalled()
    })
})