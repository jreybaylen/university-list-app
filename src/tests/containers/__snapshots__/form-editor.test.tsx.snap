// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormEditor /> Component Should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="padding: 30px; width: 90%; max-width: 400px; border-radius: 5px; margin: 8% auto 2%; box-shadow: rgb(0 0 0 / 20%) 0px 10px 15px 0px;"
      >
        <form
          name="TestEditor"
        >
          <h1
            style="font-weight: 500; margin-bottom: 30px;"
          >
            Test Form
          </h1>
          <div
            style="margin-bottom: 15px;"
          >
            <button
              style="outline: 0; cursor: pointer; width: 100%; padding: 10px 15px; border-top-right-radius: 25px; border-bottom-right-radius: 25px; text-transform: capitalize; color: rgb(255, 255, 255); border-radius: 25px; background: rgb(26, 115, 232);"
              type="submit"
            >
              Test Button
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="padding: 30px; width: 90%; max-width: 400px; border-radius: 5px; margin: 8% auto 2%; box-shadow: rgb(0 0 0 / 20%) 0px 10px 15px 0px;"
    >
      <form
        name="TestEditor"
      >
        <h1
          style="font-weight: 500; margin-bottom: 30px;"
        >
          Test Form
        </h1>
        <div
          style="margin-bottom: 15px;"
        >
          <button
            style="outline: 0; cursor: pointer; width: 100%; padding: 10px 15px; border-top-right-radius: 25px; border-bottom-right-radius: 25px; text-transform: capitalize; color: rgb(255, 255, 255); border-radius: 25px; background: rgb(26, 115, 232);"
            type="submit"
          >
            Test Button
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
